### DISCUSSION TITLE
examples not using next

### DISCUSSION THREAD

**Discussion by jlbachiochi** [pschatzmann/arduino-audio-tools#1509] [Q&A] [answered]
Most of the examples seem to be from a players point of view using next/previous to select an SD card file .  I am using my ESP32 as a source with an SD card and MP3 files I want to play on a BT speaker. Does anyone have an example of choosing a filename to play, then calling a 'play music file' routine that returns after the file has played.

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you have a look at the [Audio Player wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class)!
You could provide  your own AudioSource or use the AudioSourceCallback: this way you can basically implement any logic you want.

**Comment by jlbachiochi** [comment]
Thanks for a quick reply.  I don't think I'm trying to do anything too fancy.  Sd card files are all mp3s.  I test for all files on the inserted card.  Enable Wifi to sync time, then disable.  Setup finishes with connecting to a BT speaker.  In loop I watch for new seconds and call check for events if its a new minute.  the remainder of the loop watches player.copy() to determine if the any file is playing.
The event routine watches for quarter hours and hours to sets myMusic to a Westminster Chime mp3 file.  I'm really unsure on how to handle playing and if I need to do anything on file complete.  Attached are these areas.

void loop() 
{

    mySeconds = timeClient.getSeconds();
    if(mySeconds != lastSeconds)
    {
        newSecond();
        lastSeconds = mySeconds;
        if(mySeconds == 0)
        {
            checkForEvent();
        }
    }
    if(player.copy())
    {
        if(music)
        {
            // no change
        }
        else
        {
            music = true;
            Serial.println("Music started");
        }
    }
     else
    {
        if(!music)
        {
           // no change
        }
        else
        {
            music = false;
            player.end();
            Serial.println("Music stopped");
        }
     }
}

void playMusic()
{

    player.setAutoNext(false); 
    player.setVolume(0.1); 
    player.setPath(myMusic);
    player.begin(); 
}

The files play a few times and then they either are garbled (stuttered) or every thing freezes altogether.

**Reply by pschatzmann** [reply]
Please describe what you want to do ? 
I don't see any relation to your initial question and  I don't think that your if-then-else mess is leading anywhere...

**Comment by jlbachiochi** [comment]
The project steams mp3 files of the Westminster chime including hours gongs every hour and Westminster quarters on the quarter hour.  SD card contains 24 hour mp3 files and 3 quarter hour mp3 files.

**Comment by jlbachiochi** [comment]
The checkForEvents function assigns the appropriate file at the appropriate time to myMusic and calls playMusic.

**Reply by pschatzmann** [reply]
The AudioPlayer is there to sequentially play music files: I would not abuse this!

Try to implement first a clean logic which calls an empty play(File) method when it is needed and test this first!
Make sure and test that the play method is only called once, when it is needed.

After this is working, implementing the play method will be trivial: Just call copier.begin() with the target and source and then copier.copyAll();

Here is a simple sketch to get to started with the [relevant classes](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-sd_mp3-i2s/streams-sd_mp3-i2s.ino).

**Comment by jlbachiochi** [comment]
If I understand you correctly, the player is not the right tool to use for this.  I should be using using the stream function.  Looking at your suggestion I would bsically have to change the I2SStream to A2DPStream.

**Comment by pschatzmann** [comment]
I would test this first with another output class: e.g AnalogAudioStream or CSVStream. and do this only at the last step.
A2DP can be quite challenging! 

I expect that the playAll() will not work with A2DP, so you might need to design for using the copier.copy() in the loop like in the example.

**Comment by jlbachiochi** [comment]
I didn't make much headway by moving to the Stream class. The app makes connection with a BT speaker and waits for the quarterhour to play an Westminster chime (MP3).  The connection remains open while waiting for the next quarter hour and it plays again.  This goes on for a number of quarter hours and then the audio degrades.  Eventually it fails to play anymore and locks up.  I got this message one time...
libhelix - allocateation failed for 8708 bytes
It looks like this might be a memory issue.

**Comment by pschatzmann** [comment]
You obviously still have some bug in your code! 
Did you eventually forget to close your files ?

**Comment by jlbachiochi** [comment]
Please excuse my ignorance.  None of the examples show closing any file.
I have 3 begins...
in setup()
A2DPStream - opens BT connection
EncodedAudioStream - opens decoder
in loop()
StreamCopy - plays file
they are opened in that order
not sure where the allocation error is coming from
Since I want the connection to remain open and the same decoder always used on the MP3 files, are you suggesting after playing a file, it needs to be closed?
